File: leetcode/algorithms/121_best_time_to_buy_and_sell_stock/main.py
from typing import List


class BestTimeToBuyAndSellStock:
    """
    # Greedy
    #
    # Complexities:
    #   N - Size of `prices`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(1)
    """
    def maxProfit(self, prices: List[int]) -> int:
        minPrice: int = 100000
        maxProfit: int = 0

        for price in prices:
            if price < minPrice:
                minPrice = price

            profit: int = price - minPrice
            if profit > maxProfit:
                maxProfit = profit

        return maxProfit


    # Solution
    """
    # Greedy
    #
    # Complexities:
    #   N - Size of `prices`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(1)
    """
    def solution(self, prices: List[int]) -> int:
        min_price = float('inf')
        max_profit = 0

        for price in prices:
            if price < min_price:
                min_price = price
            elif price - min_price > max_profit:
                max_profit = price - min_price

        return max_profit

File: leetcode/algorithms/121_best_time_to_buy_and_sell_stock/main.dart
import 'dart:math';

class BestTimeToBuyAndSellStock {
  /**
   * Greedy
   *
   * Complexities:
   *   N - Size of `prices`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(1)
   */
  int maxProfit(List<int> prices) {
    int minPrice = 100000;
    int maxProfit = 0;

    for (final int price in prices) {
      if (price < minPrice) {
        minPrice = price;
      }

      int profit = price - minPrice;
      if (profit > maxProfit) {
        maxProfit = profit;
      }
    }

    return maxProfit;
  }


  // Solution
  /**
   * Greedy
   *
   * Complexities:
   *   N - Size of `prices`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(1)
   */
  int solution(List<int> prices) {
    if (prices.isEmpty) {
      return 0;
    }

    int minPrice = prices[0];
    int maxProfit = 0;

    for (int i = 1; i < prices.length; i++) {
      int currentPrice = prices[i];

      maxProfit = max(maxProfit, currentPrice - minPrice);

      minPrice = min(minPrice, currentPrice);
    }

    return maxProfit;
  }
}
